src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -239,11 +316,21 @@ private CSharpSyntaxNode ParseExpressionOrPattern(bool forCase)
                     return this.ParseIdentifierName(ErrorCode.ERR_MissingArgument);
             }
 
+            if (tk == SyntaxKind.IdentifierToken && this.CurrentToken.Text == "_")
+            {
+                // In a pattern, we reserve `_` as a wildcard. It cannot be used (with that spelling) as the
+                // type of a declaration or recursive pattern, nor as a type in an in-type expression starting
+                // in C# 7. The binder will give a diagnostic if
+                // there is a usable symbol in scope by that name. You can always escape it, using `@_`.
+                // TODO(patterns2): Should we use the "contextual keyword" infrastructure for this?
```

Comment:
>TODO [](start = 19, length = 4)

PROTOTYPE?

Reply:
This has already been recorded in [https://github.com/dotnet/roslyn/issues/23222](https://github.com/dotnet/roslyn/issues/23222).

---
In reply to: [156498946](https://github.com/dotnet/roslyn/pull/23209#discussion_r156498946) [](ancestors = 156498946)